### PR TITLE
WIP Reclaimer

### PR DESCRIPTION
**Pull request by brossetti1**
WIP branch for Reclaimer https://github.com/Geo-Web-Project/core-contracts/issues/51

### CODE REVIEW

**Comment by brossetti1** [comment]
important aspected of this so far are:

1) [getting the `startTime` for the reclaimer](https://github.com/Geo-Web-Project/core-contracts/pull/55/files#diff-910f6e5d941506109ad63706f1fab09b85820173ce88001529ceadd8f9e94324R310-R319) - I made this helper on the superapp thinking it marks the `timestamp` when the currentBid owner defaulted on their payment.. 

2) not sure what to do in the [`claim` method since](https://github.com/Geo-Web-Project/core-contracts/pull/55/files#diff-c6590f2e62f39b7459331f677fff154d7acf66d3394cf7cdaed5569921e15368R52) it looks like the superapp already takes care of `license.safeTransfer` to the new `bidder`

3) [claimPrice claculation](https://github.com/Geo-Web-Project/core-contracts/pull/55/files#diff-c6590f2e62f39b7459331f677fff154d7acf66d3394cf7cdaed5569921e15368R64-R84) - current logic for the `claimPrice` calculation

File: contracts/AuctionSuperApp.sol [hunk 1]
```diff
@@ -307,6 +307,16 @@ contract AuctionSuperApp is SuperAppBase, AccessControlEnumerable, Pausable {
         return bid.contributionRate;
     }
 
+    /**
+     * @notice get the start time of the reclaimer bid to calculate reclaimer.claimPrice
+     * @param id License id
+     * @return reclaim start time
+     */
+    function getReclaimerStartTime(uint256 id) external view returns (uint256) {
```
Comment (codynhat) [inline contracts/AuctionSuperApp.sol:315]: I think this should be the latest of `lastUserDeletion[owner]` and `currentOwnerBid[id].timestamp`.

- `currentOwnerBid[id].timestamp` will be updated if an owner lowers their payment to 0 but still has other parcels
- `lastUserDeletion[owner]` will be updated if an owner deletes their flow, which lowers their payment to 0 for all parcels they own

In either case, reclaiming will only be possible if `ownerBidContributionRate() == 0`. The super app checks for this before calling the reclaimer.

File: contracts/ReClaimer.sol [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        if (license.ownerOf(licenseId) == address(0x0)) {
+            revert("ReClaimer: Cannot reclaim non-existent license");
+        }
+
+		uint256 licenseId  = abi.decode(claimData, (uint256));
+        uint256 startTime = auctionApp.getReclaimerStartTime(licenseId);
+        require(startTime != 0, "parcel is not for reclamation");
+		require(block.timestamp > startTime, "no relcaimer auction on this license");
+
+        // TODO implement reclaim?
```
Comment (codynhat) [inline contracts/ReClaimer.sol:52]: You're right, the super app already does the transfer. Maybe we should just move the transfer to happen here instead?

**Comment by brossetti1** [comment]
@codynhat i updated based on your feedback - wondering if you aggree with these changes:

[get correct reclaim start time](https://github.com/Geo-Web-Project/core-contracts/pull/55/files#diff-910f6e5d941506109ad63706f1fab09b85820173ce88001529ceadd8f9e94324R777-R793)
[Update `claimdata` for calculating reclaim price](https://github.com/Geo-Web-Project/core-contracts/pull/55/files#diff-c6590f2e62f39b7459331f677fff154d7acf66d3394cf7cdaed5569921e15368R69-R91)

if you agree with these changes - ill move on to writing tests

**Review by codynhat — APPROVED** [review]

**Pull request merged**